services/scan/src/precinct_scanner_app.ts

```diff
@@ -0,0 +1,454 @@
+  app.post('/scan/hmpb/doneTemplates', async (_request, response) => {
+    await configureMachine(machine, workspace);
+    response.json({ status: 'ok' });
+  });
+
+  app.post<NoParams, Scan.ExportResponse, Scan.ExportRequest>(
+    '/scan/export',
+    (_request, response) => {
+      const outputStream = new streams.WritableStream();
+      store.exportCvrs(outputStream);
+      const cvrs = outputStream.toString();
+      store.setCvrsAsBackedUp();
+      response.set('Content-Type', 'text/plain; charset=utf-8');
+      response.send(cvrs);
+    }
+  );
+
+  app.get('/scan/backup', (_request, response) => {
+    console.log('backup');
+    const electionDefinition = store.getElectionDefinition();
+
+    if (!electionDefinition) {
+      response.status(500).json({
+        errors: [
+          {
+            type: 'unconfigured',
+            message: 'cannot backup an unconfigured server',
+          },
+        ],
+      });
+      return;
+    }
+
+    response
+      .header('Content-Type', 'application/zip')
+      .header(
+        'Content-Disposition',
+        `attachment; filename="election-${electionDefinition.electionHash.slice(
+          0,
+          10
+        )}-${new Date()
+          .toISOString()
+          .replace(/[^-a-z0-9]+/gi, '-')}-backup.zip"`
+      )
+      .flushHeaders();
+
+    backup(store)
+      .on('error', (error: Error) => {
+        // debug('backup error: %s', error.stack);
+        response.status(500).json({
+          errors: [
+            {
+              type: 'error',
+              message: error.toString(),
+            },
+          ],
+        });
+      })
+      .on('end', () => {
+        store.setBatchesAsBackedUp();
+      })
+      .pipe(response);
+  });
+
+  app.get<NoParams, Scan.GetPrecinctScannerStatusResponse>(
+    '/scanner/status',
+    (_request, response) => {
+      const machineStatus = machine.status();
+      const batches = store.batchStatus();
+      const ballotsCounted = sum(batches.map((batch) => batch.count));
+      const canUnconfigure = store.getCanUnconfigure();
+      response.json({
+        ...machineStatus,
+        ballotsCounted,
+        canUnconfigure,
+      });
+    }
+  );
+
+  app.post<NoParams, OkResponse>('/scanner/scan', (_request, response) => {
+    machine.scan();
+    response.json({ status: 'ok' });
+  });
+
+  app.post<NoParams, OkResponse>('/scanner/accept', (_request, response) => {
+    machine.accept();
+    response.json({ status: 'ok' });
+  });
+
+  app.post<NoParams, OkResponse>('/scanner/return', (_request, response) => {
+    machine.return();
+    response.json({ status: 'ok' });
+  });
+
+  app.get('/*', (request, response) => {
+    const url = new URL(`http://${request.get('host')}${request.originalUrl}`);
+    url.port = '3000';
+    response.redirect(301, url.toString());
+  });
+
+  // Reload configuration from store on startup
+  if (store.getElectionDefinition()) {
+    void configureMachine(machine, workspace);
+  }
```

Comment:
@eventualbuddha does this seem right?

Reply:
Yeah, I think so. The thing to be careful of with `void …` to essentially "background" a `Promise`-returning function is that you should be quite sure that it won't throw under normal circumstances. In future (unspecified) versions of NodeJS an unhandled promise rejection will crash the process.

Will this get called again somehow if the machine doesn't boot up already configured?

Reply:
Oh ok interesting. Any thoughts on how to make it safe but non-blocking?

This function also gets called in the `doneTemplates` endpoint (which I think signifies the end of configuration)

Reply:
Mostly just make sure it handles anything that might be thrown by things it calls, except for `assert`s. Probably means you'll want to think about what this should do if loading configuration fails. Have some sort of configuration error state?